staging/vhost-device-gpu/src/protocol.rs

```diff
@@ -0,0 +1,1453 @@
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_ERR_INVALID_SCANOUT_ID as VIRTIO_GPU_RESP_ERR_INVALID_SCANOUT_ID,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_ERR_OUT_OF_MEMORY as VIRTIO_GPU_RESP_ERR_OUT_OF_MEMORY,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_ERR_UNSPEC as VIRTIO_GPU_RESP_ERR_UNSPEC,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_CAPSET as VIRTIO_GPU_RESP_OK_CAPSET,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_CAPSET_INFO as VIRTIO_GPU_RESP_OK_CAPSET_INFO,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_DISPLAY_INFO as VIRTIO_GPU_RESP_OK_DISPLAY_INFO,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_EDID as VIRTIO_GPU_RESP_OK_EDID,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_MAP_INFO as VIRTIO_GPU_RESP_OK_MAP_INFO,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_NODATA as VIRTIO_GPU_RESP_OK_NODATA,
+    virtio_gpu_ctrl_type_VIRTIO_GPU_RESP_OK_RESOURCE_UUID as VIRTIO_GPU_RESP_OK_RESOURCE_UUID,
+};
+use virtio_queue::{Reader, Writer};
+use vm_memory::{ByteValued, GuestAddress, Le32};
+use zerocopy::{AsBytes, FromBytes};
+
+use crate::device::{self, Error};
+
+pub const QUEUE_SIZE: usize = 1024;
+pub const NUM_QUEUES: usize = 2;
+
+pub const CONTROL_QUEUE: u16 = 0;
+pub const CURSOR_QUEUE: u16 = 1;
+pub const POLL_EVENT: u16 = 3;
+
+pub const VIRTIO_GPU_MAX_SCANOUTS: u32 = 16;
+
+/// `CHROMIUM(b/277982577)` success responses
+pub const VIRTIO_GPU_RESP_OK_RESOURCE_PLANE_INFO: u32 = 0x11FF;
+
+/// Create a OS-specific handle from guest memory (not upstreamed).
+pub const VIRTIO_GPU_BLOB_FLAG_CREATE_GUEST_HANDLE: u32 = 0x0008;
+
+pub const VIRTIO_GPU_FLAG_FENCE: u32 = 1 << 0;
+pub const VIRTIO_GPU_FLAG_INFO_RING_IDX: u32 = 1 << 1;
+
+/// Virtio Gpu Configuration
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+#[repr(C)]
+pub struct VirtioGpuConfig {
+    /// Signals pending events to the driver
+    pub events_read: Le32,
+    /// Clears pending events in the device
+    pub events_clear: Le32,
+    /// Maximum number of scanouts supported by the device
+    pub num_scanouts: Le32,
+    /// Maximum number of capability sets supported by the device
+    pub num_capsets: Le32,
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioGpuConfig {}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidCommandType(u32);
+
+impl std::fmt::Display for InvalidCommandType {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(fmt, "Invalid command type {}", self.0)
+    }
+}
+
+impl From<InvalidCommandType> for crate::device::Error {
+    fn from(val: InvalidCommandType) -> Self {
+        Self::InvalidCommandType(val.0)
+    }
+}
+
+impl std::error::Error for InvalidCommandType {}
+
+#[derive(Copy, Clone, Debug, Default, AsBytes, FromBytes, PartialEq, Eq)]
+#[repr(C)]
+pub struct virtio_gpu_ctrl_hdr {
+    pub type_: u32,
```

Comment:
Should these virtio_gpu types be using little endian or is there a reason why we use host endianess here? (I took a skim at the code and it wasn't immediately obvious)

Reply:
I missed that it should be little-endian. Thanks for pointing that out!
these structs should follow the VirtIO specification.

Reply:
Not just this one btw, do a search for `u32` and `u64` in this module to find other ones.